src/network/routing/onion_request_router.cpp

```diff
@@ -0,0 +1,1218 @@
+            case RequestCategory::download: return PathSelectionBehaviour::new_or_least_busy;
+        }
+        return PathSelectionBehaviour::random;
+    }
+
+    std::vector<service_node> extract_nodes(
+            const std::unordered_map<RequestCategory, std::vector<OnionPath>>& paths,
+            const std::unordered_map<std::string, std::vector<service_node>>& pending_paths) {
+        std::vector<service_node> all_used_nodes;
+
+        for (const auto& [pt, path_list] : paths)
+            for (const auto& p : path_list)
+                all_used_nodes.insert(all_used_nodes.end(), p.nodes.begin(), p.nodes.end());
+
+        for (const auto& [pid, nodes] : pending_paths)
+            all_used_nodes.insert(all_used_nodes.end(), nodes.begin(), nodes.end());
+
+        return all_used_nodes;
+    }
+}  // namespace
+
+std::string OnionPath::to_string() const {
+    std::vector<std::string> node_descriptions;
+    std::transform(
+            nodes.begin(),
+            nodes.end(),
+            std::back_inserter(node_descriptions),
+            [](const service_node& node) { return node.to_string(); });
+
+    return "{}"_format(fmt::join(node_descriptions, ", "));
+}
+
+OnionRequestRouter::OnionRequestRouter(
+        config::OnionRequestRouterConfig config,
+        std::shared_ptr<oxen::quic::Loop> loop,
+        std::weak_ptr<SnodePool> snode_pool,
+        std::weak_ptr<ITransport> transport) :
+        _config{std::move(config)}, _loop{loop}, _snode_pool{snode_pool}, _transport{transport} {
+    log::trace(cat, "[OnionRequestRouter] Initializing.");
+
+    _request_queues[RequestCategory::standard] =
+            std::make_shared<detail::RequestQueue>(loop, _config.request_timeout_check_frequency);
+    _request_queues[RequestCategory::upload] =
+            std::make_shared<detail::RequestQueue>(loop, _config.request_timeout_check_frequency);
+    _request_queues[RequestCategory::download] =
+            std::make_shared<detail::RequestQueue>(loop, _config.request_timeout_check_frequency);
+
+    _loop->call_soon([this] {
+        auto snode_pool = _snode_pool.lock();
+        if (!snode_pool) {
+            log::critical(
+                    cat, "[OnionRequestRouter] SnodePool was destroyed, cannot setup router.");
+            return;
+        }
+
+        if (snode_pool->size() == 0)
+            snode_pool->refresh_if_needed({}, [weak_self = weak_from_this()] {
+                if (auto self = weak_self.lock())
+                    self->_loop->call([weak_self] {
+                        if (auto self = weak_self.lock())
+                            self->_finish_setup();
+                    });
+            });
+        else
+            _finish_setup();
+    });
+}
+
+OnionRequestRouter::~OnionRequestRouter() {
+    // Use 'call_get' to force this to be synchronous
+    if (_loop)
+        _loop->call_get([this] { _close_connections(); });
+    log::debug(cat, "[OnionRequestRouter] Destroyed.");
```

Comment:
All of these `[OnionRequestRouter]` embedded strings feels like a duplication of what the log::Cat is meant to be.  Can we change this around to `auto cat = oxen::log::Cat("OnionRequestRouter");` and then drop all of these redundant message prefixes?